src/item_pocket.cpp

```diff
@@ -0,0 +1,592 @@
+                                   string_format( "<num> %s", volume_units_abbr() ), f,
+                                   converted_volume );
+            } else {
+                info.emplace_back( "DESCRIPTION", contents_item.display_name() );
+            }
+        }
+    }
+}
+
+ret_val<item_pocket::contain_code> item_pocket::can_contain( const item &it ) const
+{
+    // legacy container must be added to explicitly
+    if( type == pocket_type::LEGACY_CONTAINER ) {
+        return ret_val<item_pocket::contain_code>::make_failure( contain_code::ERR_LEGACY_CONTAINER );
+    }
+    if( it.made_of( phase_id::LIQUID ) && !watertight ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_LIQUID, _( "can't contain liquid" ) );
+    }
+    if( it.made_of( phase_id::GAS ) && !gastight ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_GAS, _( "can't contain gas" ) );
+    }
+    if( it.volume() < min_item_volume ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_TOO_SMALL, _( "item is too small" ) );
+    }
+    if( it.weight() > max_contains_weight ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_TOO_HEAVY, _( "item is too heavy" ) );
+    }
+    if( it.weight() > remaining_weight() ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_CANNOT_SUPPORT, _( "pocket is holding too much weight" ) );
+    }
+    if( !flag_restriction.empty() && !it.has_any_flag( flag_restriction ) ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_FLAG, _( "item does not have correct flag" ) );
+    }
+    if( it.volume() > max_contains_volume ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_TOO_BIG, _( "item too big" ) );
+    }
+    if( it.volume() > remaining_volume() ) {
+        return ret_val<item_pocket::contain_code>::make_failure(
+                   contain_code::ERR_NO_SPACE, _( "not enough space" ) );
+    }
+    return ret_val<item_pocket::contain_code>::make_success();
+}
+
+cata::optional<item> item_pocket::remove_item( const item &it )
+{
+    item ret( it );
+    const size_t sz = contents.size();
+    contents.remove_if( [&it]( const item & rhs ) {
+        return &rhs == &it;
+    } );
+    if( sz == contents.size() ) {
+        return cata::nullopt;
+    } else {
+        return ret;
+    }
+}
+
+cata::optional<item> item_pocket::remove_item( const item_location &it )
+{
+    if( !it ) {
+        return cata::nullopt;
+    }
+    return remove_item( *it );
+}
+
+bool item_pocket::spill_contents( const tripoint &pos )
+{
+    for( item &it : contents ) {
+        g->m.add_item_or_charges( pos, it );
```

Comment:
Should this be potentially spilling into a vehicle rather than directly onto the map?  (If this is just imitating existing behaviour, then don't worry about it).

Reply:
this is just imitating current behavior.